[AppService] az webapp deployment github-actions add/remove: new commands

calvinsID opened the pull request:
**Description**<!--Mandatory-->
New commands for adding/removing github actions to a webapp
- `az webapp deployment github-actions add`
- `az webapp deployment github-actions remove`

**Testing Guide**
Example usage:

- Add:

![image](https://user-images.githubusercontent.com/14339314/119858624-7f659600-bec9-11eb-92ab-a48f927932af.png)

Navigate to https://github.com/login/device to enter code
![image](https://user-images.githubusercontent.com/14339314/119858808-a91ebd00-bec9-11eb-9508-e88166c56087.png)

Authorize after entering in code
![image](https://user-images.githubusercontent.com/14339314/120008207-92db3480-bf8f-11eb-8905-bc41ca5cac0c.png)

- Remove:
![image](https://user-images.githubusercontent.com/14339314/119858991-d0758a00-bec9-11eb-90a8-039f47dc2328.png)

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

JasonFreeberg commented on src/azure-cli/azure/cli/command_modules/appservice/custom.py:
```diff
@@ -4461,3 +4465,449 @@ def delete_function_key(cmd, resource_group_name, name, key_name, function_name=
+    import uuid
+    guid = str(uuid.uuid4()).replace('-', '')
+    publish_profile_name = "AzureAppService_PublishProfile_{}".format(guid)
+    logger.warning('Fill workflow template with name: {}, branch: {}, version: {}, slot: {}'.format(name, branch, github_actions_version, slot if slot else 'production'))
```
Fill --> Filling

JasonFreeberg commented on src/azure-cli/azure/cli/command_modules/appservice/custom.py:
```diff
@@ -4461,3 +4465,449 @@ def delete_function_key(cmd, resource_group_name, name, key_name, function_name=
+                                content=completed_workflow_file, branch=branch)
+
+    # Add publish profile to GitHub
+    logger.warning('Add publish profile to GitHub')
```
Add --> Adding

JasonFreeberg reviewed (changes requested):
I added two comments about the logging to use a consistent verb tense. 

Can someone explain what's happening with the browser auth? From what I see, the CLI command will open the browser to start the auth flow. Once the user does that, is a PAT token created and placed on the user's dev box? Just trying to understand the auth flow better.

yonzhan commented:
@Juliehzl to review

calvinsID commented:
> I added two comments about the logging to use a consistent verb tense.
> 
> Can someone explain what's happening with the browser auth? From what I see, the CLI command will open the browser to start the auth flow. Once the user does that, is a PAT token created and placed on the user's dev box? Just trying to understand the auth flow better.

Thanks for the review Jason! I've updated the description with a few more details, the CLI command will tell you to navigate to the url "github.com/login/device" and enter in the code to authorize us. We will poll and wait up to 14 minutes for the user to do this, and once the user does, we will receive the PAT token from polling and use that token in the command. The token won't be stored anywhere so the user will have to do this process everytime unless they supply the PAT token. We can add a command to just retrieve the PAT token in the future, but I wasn't sure of where to put/how to name the command

Juliehzl commented on src/azure-cli/azure/cli/command_modules/appservice/_help.py:
```diff
@@ -1715,6 +1715,35 @@
+
+helps['webapp deployment github-actions add'] = """
+    type: command
+    short-summary: Adds a GitHub Actions workflow file to the specified repository. The workflow will build and deploy your app to the specified webapp.
```
```suggestion
    short-summary: Add a GitHub Actions workflow file to the specified repository. The workflow will build and deploy your app to the specified webapp.
```

Juliehzl commented on src/azure-cli/azure/cli/command_modules/appservice/_help.py:
```diff
@@ -1715,6 +1715,35 @@
+
+helps['webapp deployment github-actions remove'] = """
+    type: command
+    short-summary: Removes and disconnects the GitHub Actions workflow file from the specified repository.
```
```suggestion
    short-summary: Remove and disconnects the GitHub Actions workflow file from the specified repository.
```

Juliehzl commented on src/azure-cli/azure/cli/command_modules/appservice/_params.py:
```diff
@@ -410,6 +410,21 @@ def load_arguments(self, _):
+        c.argument('slot', options_list=['--slot', '-s'], help='The name of the slot. Default to the production slot if not specified.')
+        c.argument('branch', options_list=['--branch', '-b'], help='The branch to which the workflow file will be added. Defaults to "master" if not specified.')
+        c.argument('force', options_list=['--force', '-f'], help='When true, the command will overwrite any workflow file with a conflicting name.')
+        c.argument('login_with_github', help='Interactively log in with Github to retrieve the Personal Access Token')
```
If this flag will take effect when specified, recommend to use `action='store_true'`.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/appservice/_params.py:
```diff
@@ -410,6 +410,21 @@ def load_arguments(self, _):
+
+    with self.argument_context('webapp deployment github-actions add')as c:
+        c.argument('runtime', options_list=['--runtime', '-r'], help='Canonicalized web runtime in the format of Framework|Version, e.g. "PHP|5.6". Use "az webapp list-runtimes" for available list.')
+        c.argument('force', options_list=['--force', '-f'], help='When true, the command will overwrite any workflow file with a conflicting name.')
```
If you want to make it take effect when specified, use `action='store_true`

Juliehzl commented on src/azure-cli/azure/cli/command_modules/appservice/commands.py:
```diff
@@ -247,6 +247,10 @@ def load_command_table(self, _):
         g.custom_command('config', 'enable_cd')
         g.custom_command('show-cd-url', 'show_container_cd_url')
 
+    with self.command_group('webapp deployment github-actions') as g:
```
what about adding preview tag for one or two sprint to test the functionality for new commands?

Juliehzl commented on src/azure-cli/setup.py:
```diff
@@ -139,6 +139,7 @@
     'jsondiff==1.2.0',
     'packaging~=20.9',
     'PyGithub==1.38',
+    'PyNaCl==1.4.0',
```
please also add in requirementxxxx files.

calvinsID replied:
It's already added in requirementsxxx files, so I just added here as well

jiasli replied:
Why do we need to pin `PyNaCL`?

calvinsID replied:
unpinned

Juliehzl commented:
Hi @calvinsID, before merge, please also get approval from other code owners and @JasonFreeberg as he is requesting change,

jiasli commented on src/azure-cli/setup.py:
```diff
@@ -140,6 +140,7 @@
     'jsondiff==1.2.0',
     'packaging~=20.9',
     'PyGithub==1.38',
```
Could you unpin `PyGithub` (#18548, #17826)?

calvinsID replied:
Unpinned in `setup.py`. Wondering why in requirements.txt all are pinned?

jiasli replied:
`requirements.txt` is used for building [official packages](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli) like msi, deb, rpm. It is usually not possible to be used together with other python libraries in the same environment. That's why we want to pin the version. 

Also, you may check the `pip` doc on `requirements.txt`: https://pip.pypa.io/en/stable/user_guide/#requirements-files

> Requirements files are used to hold the result from pip freeze for the purpose of achieving repeatable installations. In this case, your requirement file contains a pinned version of everything that was installed when pip freeze was run.

panchagnula reviewed (approved):
shipIt

Juliehzl commented on src/azure-cli/azure/cli/command_modules/appservice/_help.py:
```diff
@@ -1715,6 +1715,35 @@
+"""
+
+helps['webapp deployment github-actions add'] = """
+    type: command
+    short-summary: Add a GitHub Actions workflow file to the specified repository. The workflow will build and deploy your app to the specified webapp.
+    examples:
+      - name: Add GitHub Actions to a specified repository, providing personal access token
+        text: >
+            az webapp deployment github-actions add --repo "githubUser/githubRepo" -g MyResourceGroup -n MyWebapp --token MyPersonalAccessToken
+      - name: Add GitHub Actions to a specified repository, using interactive method of retrieving personal access token
+        text: >
+            az webapp deployment github-actions add --repo "githubUser/githubRepo" -g MyResourceGroup -n MyWebapp --login-with-github
```
it would be better if you can remove the spaces here to align with other help entries

Juliehzl commented on src/azure-cli/azure/cli/command_modules/appservice/_help.py:
```diff
@@ -1715,6 +1715,35 @@
+"""
+
+helps['webapp deployment github-actions remove'] = """
+    type: command
+    short-summary: Remove and disconnect the GitHub Actions workflow file from the specified repository.
+    examples:
+      - name: Remove GitHub Actions from a specified repository, providing personal access token
+        text: >
+            az webapp deployment github-actions remove --repo "githubUser/githubRepo" -g MyResourceGroup -n MyWebapp --token MyPersonalAccessToken
+      - name: Remove GitHub Actions from a specified repository, using interactive method of retrieving personal access token
+        text: >
+            az webapp deployment github-actions remove --repo "githubUser/githubRepo" -g MyResourceGroup -n MyWebapp --login-with-github
```
the same

Juliehzl commented on src/azure-cli/setup.py:
```diff
@@ -139,7 +139,8 @@
     'jsondiff==1.2.0',
     'packaging~=20.9',
-    'PyGithub==1.38',
+    'PyGithub~=1.38',
+    'PyNaCl~=1.4.0',
```
you also need to add dependencies in requirementxxx.txt

calvinsID replied:
These already existed in requirementxxx.txt

Juliehzl commented on src/azure-cli/azure/cli/command_modules/appservice/utils.py:
```diff
@@ -75,3 +75,11 @@ def call(*args, **kwargs):
+
+def raise_missing_token_suggestion():
+    pat_documentation = "https://help.github.com/en/articles/creating-a-personal-access-token-for-the-command-line"
+    raise CLIError("GitHub access token is required to authenticate to your repositories. "
```
recommend to use the exception type in azclierror, e.g.
https://github.com/Azure/azure-cli/blob/3eea7fd773722b7a3c40b8115f29a3bad39b70bc/src/azure-cli-core/azure/cli/core/azclierror.py#L142-L144

panchagnula commented:
@Juliehzl can you review the lashes updates on this one & approve & merge?
